###  <font color='blue'>Background information about Python etc. for the homework 2 assignment. 

If you are not familiar with Jupyter notebooks, we recommend that you read [this tutorial](https://realpython.com/jupyter-notebook-introduction/) or watch [this short video tutorial](https://www.youtube.com/watch?v=HW29067qVWk) (you only need to watch the first 18min). At the very least make sure you understand how to create text and code cells, how to move them around, and how to run code cells. 

### <font color='blue'>Euler's number $e$:</font> 

Expression for the Poisson probability distribution includes [Euler's number] $e$. It arises in various mathematical calculations, including the basic calculation of compound interest - see Appendix B, p. 113, of [David Morin's book](https://en.wikipedia.org/wiki/E_(mathematical_constant))).

For example, the code below shows how $(1+1/n)^n$ converges to $e$ as $n$ increases

In [ ]:
from math import e # math is a Python module containing various mathematical modules

for n in range (1000000, 10000000, 1000000):
    print(n, (1+1/n)**n, e)

1000000 2.7182804690957534 2.718281828459045
2000000 2.7182811492688552 2.718281828459045
3000000 2.7182813757921678 2.718281828459045
4000000 2.7182814890538505 2.718281828459045
5000000 2.718281555200129 2.718281828459045
6000000 2.71828160231555 2.718281828459045
7000000 2.7182816328653 2.718281828459045
8000000 2.7182816565320875 2.718281828459045
9000000 2.7182816796340985 2.718281828459045


You can either set e yourself: 

    e = 2.718281828459045

or "import" it from the <tt>math</tt> module as shown above. 

### <font color='blue'>While loops

Allow loops that iterate until some condition is satisfied and stop when it is violated. 

    while condition is true: 
        do something
    else:
        do something else to end the loop
    
The last <tt>else:</tt> branch is usually omitted, but can be useful sometimes to perform a useful information to indicate that loop was not entered (because its condition is violated right away, or that loop finished). 

In [ ]:
x = 5
while x > 0:
    x -= 1
    print(x)

4
3
2
1
0


As described in [W3S section on while loops](https://www.w3schools.com/python/python_while_loops.asp) one can use <tt>break</tt> and <tt>continue</tt> statements with loops. 

<tt>break</tt> statement breaks loop iterations even if condition in the while loop is still true or the range of indices in <tt>range</tt> is not yet exhausted. It will transfer execution to the code immediately following the loop.

In [ ]:
x = 10
while x: # while x>0
    x = x - 1 # countdown!
    print(x)
else: 
    print('countdown finished normally - blastoff!')

print('did something break?')

9
8
7
6
5
4
3
2
1
0
countdown finished normally - blastoff!
did something break?


We can see that in the normal execution without break, code executed <tt>else:</tt> branch and printed its print statement, but with a <tt>break statement</tt> below execution goes directly to the print statement outside the loop, x only gets down to 5 and <tt>else</tt> branch is not executed. 

In [ ]:
x = 10
while x: # while x>0
    x = x - 1 # countdown!
    if x == 5: 
        break
    print(x)
else: 
    print('countdown finished normally - blastoff!')

print('did something break?')

9
8
7
6
did something break?


<tt>continue</tt> statement skips the code after it only for the *current* iteration of the loop and simply transfers execution to the next loop (instead of breaking from loop iterations altogether).. 

In the code below we can see that during iteration when x became 5, <tt>print(x)</tt> following <tt>if</tt> branch did not execute, because 5 was not printed. 

In [ ]:
x = 10
while x: # while x>0
    x = x - 1 # countdown!
    if x == 5: 
        print('hiccup at x =', x)
        continue
    print(x)
else: 
    print('countdown finished normally - blastoff!')

print('did something break?')

9
8
7
6
hiccup at x = 5
4
3
2
1
0
countdown finished normally - blastoff!
did something break?


<tt>continue</tt> and <tt>break</tt> statements can be useful in some situations, but most code can actually be written without them. For example, the code above can simply be rewritten to the same effect without continue like this:

In [ ]:
x = 10
while x: # while x>0
    x = x - 1 # countdown!
    if x != 5:
        print(x)
    else:
        print('hiccup at x =', x)
else: 
    print('countdown finished normally - blastoff!')

print('did something break?')

9
8
7
6
hiccup at x = 5
4
3
2
1
0
countdown finished normally - blastoff!
did something break?


General recommendation is to use <tt>continue</tt> and <tt>break</tt> statements sparingly because they result in jumps in code execution sequence, which sometimes can make it difficult to understand what code does and to find errors in it. 

When you feel like you should be using one of these statements, it's a good mental exercise to think about how to write code so you don't need to use them. 

Having said this, there are situations when break is useful. It can allow you to break out of a very long and computationally expensive loop without running through all of its iterations, if running them no longer makes sense. 

### <font color='blue'>Lists</font> 

Lists are a data structure consisting of ordered sequences of Python objects, and can contain objects of different type.

A gentle introduction to lists can be found at W3 School site [here](https://www.w3schools.com/python/python_lists.asp)


In [1]:
lst = [1, 2, 3, 4, 5]
lst

[1, 2, 3, 4, 5]

We can initialize lists an empty list or a list consisting of the same elements like this:

In [2]:
empty_list = [] # this defines a list with no elements

As numbers, lists can be used in logical evaluations in <tt>if</tt> or <tt>while</tt> control structures because an empty list evaluates to 0 (False) in such structures. 

In [3]:
if empty_list:
    print('this list is not empty!')
else: 
    print('the list is empty')

the list is empty


In [4]:
empty_list == False # this does not work though

False

In [5]:
n = 10
numbers = [55]*n
numbers

[55, 55, 55, 55, 55, 55, 55, 55, 55, 55]

Lists can be composed of elements of different type. 

In [6]:
lst = [1, 2., '3', 4, '5']
lst

[1, 2.0, '3', 4, '5']

individual elements (entries) of the list can be accessed by their index corresponding to their order in the list:

In [10]:
print(lst[0], lst[2], lst[-2], lst[-1])

1 3 4 5


Note that index -1 corresponds to the last element of the list, -2 to the next to last, etc. 

We can determine how many elements a list has by using built-in function <tt>len</tt>

In [8]:
len(lst)

5

After initialization we can append elements at the end of the list. <tt>append</tt> here is one of the list *methods*. 

In [9]:
x = []

for i in range(11): 
    x.append(i)

print(x)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]



**Methods associated with lists**, such as <tt>list.append()</tt>, are described [here](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists)

### <font color='blue'>Slicing operations on lists</font> 


In [ ]:
lst[::] # scan through list elements

[1, 2.0, '3', 4, '5']

In [ ]:
lst[::-1] # scan through elements in reverse order 

['5', 4, '3', 2.0, 1]

In [ ]:
lst[::2], lst[::-2] # scan with step of 2, scan with step 2 in reverse order

([1, '3', '5'], ['5', '3', 1])

In [ ]:
lst[2:] # scan starting from the 3rd element

['3', 4, '5']

In [ ]:
lst[:-2] # scan starting from the 3rd to last element

[1, 2.0, '3']

Lists can be changed at any point (they are "mutable" in programming jargon):

In [ ]:
lst[4] = 'five'
lst

[1, 2.0, '3', 4, 'five']

We can generate a list of consecutive numbers from n1 to n2-1 with some step s as follows

In [ ]:
n1, n2, s = 0, 10, 1
consecutive_numbers = list(range(n2)) # in this case n1 and s can be ommitted as 0 and 1 are their default values
print(consecutive_numbers)

n1, n2, s = 10, 100, 10
consecutive_numbers = list(range(n1, n2, s))
print(consecutive_numbers)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[10, 20, 30, 40, 50, 60, 70, 80, 90]


### <font color='blue'>List assignment using another list vs cloning a list 

There is a subtle, but important issue with assignments involving two lists. When assinment is done as <tt>list1 = list2</tt> what happens is that <tt>list1</tt> is assigned a pointer to the elements of <tt>list2</tt> in memory. Thus, both list names are now referencing the same location in memory, so if elements in one list are changed, they will change in the other. For example: 

In [ ]:
x = list(range(5))
print(x)
y = x
y[-1] = 'four'
print(x) # here we will see that x changed even though we did not explicitly change it

[0, 1, 2, 3, 4]
[0, 1, 2, 3, 'four']


If we want to create an *independent* copy of a list, we can either generate two lists independently 

In [ ]:
x = list(range(5))
print(x)
y = list(range(5))
y[-1] = 'four'
print(x) # here we will see that x changed even though we did not explicitly change it

[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4]


Or, if this is not easy, we can  ***clone*** the list as follows

In [ ]:
x = list(range(5))
print(x)
z = x[:]
z[4] = 'four'
print(x)

[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4]


**Note** that using a slicing operation on a list always creates a copy of this list in memory. Likewise, we can use slicing to copy only a subset of list elements: 

In [ ]:
z = x[1:3]
print(z)

[1, 2]


### <font color='blue'>Functions and variable "scope"

variables assigned within functions cannot be accessed outside functions, unless they are explicitly defined as <tt>global</tt>:

In [ ]:
def happy():
    h = 'happy'
    print(h)

happy()
print(h)

happy


NameError: name 'h' is not defined

In [ ]:
# but 
def happy():
    global h 
    h = 'happy'
    print(h)

happy()
print(h)

happy
happy


### <font color='blue'>Functions returning values via <tt>return</tt> statement(s)

Functions can return certain results of their internal code execution to the outside program or don't return anything, just perform some actions and exit to the main program. 

<tt>return</tt> statements allow to pass on values of variables or results of calculations within a function to outside code, without using <tt>global</tt> declaration. 

Functions that don't return any variables explicitly, implicitly return a special value <tt>None</tt> of type <tt>NoneType</tt>

In [ ]:
def return_moods():
    happy_moods = ['happy', 'exuberant', 'glad', 'exhilarated', 'on-the-roll', 'in the flow']
    neutral_moods = ['ok', 'so-so']
    unhappy_moods = ['unhappy', 'sad', 'blue']
    return happy_moods, neutral_moods, unhappy_moods  

h, n, u = return_moods()
print("happy moods: ", h)
print("neutral moods:", n)
print("unhappy moods: ",u)

happy moods:  ['happy', 'exuberant', 'glad', 'exhilarated', 'on-the-roll', 'in the flow']
neutral moods: ['ok', 'so-so']
unhappy moods:  ['unhappy', 'sad', 'blue']


When function returns multiple values, as the function above, it actually returns a tuple. This tuple is then assigned through chain assignment (also called tuple assignment): 

    h, n, u = return_moods()

In [ ]:
def happy_():
    print('Happy birthday to you!')
    
def happy_bday(name):
    if name != '': 
        happy_(); happy_()
        print('Happy birthday, dear ' + name + '!')
        happy_()
        
name = input()
res = happy_bday(name)

X
Happy birthday to you!
Happy birthday to you!
Happy birthday, dear X!
Happy birthday to you!


In [ ]:
# functions without explicit return statements 
# return a special None type value
print(res, type(res))

None <class 'NoneType'>


### <font color='blue'>Functions taking variable number of parameters 

When we want to create a function that are flexible and can take variable number of parameters, we can use *pointer* operator in front of the argument variable name, to point to the beginning of the argument sequence. 

    def func(*args): 
        for arg in args: 
            print(arg)

In [ ]:
def func(*args): 
    for arg in args: 
        print(arg, end = ' ')

print(func()) # this shows that by default functions have one dummy parameter None (of type NoneType)
print(func(1,2))
print(func(1,2,3))
print(func('one', 'two', 'three', 'four'))

None
1 2 None
1 2 3 None
one two three four None


Two ways of using args parameter.

In [ ]:
def func(x, *args):
    print(args)
    a, b, c = args
    return x

def func2(x, a, b, c):
    print(a, b, c)
    return x

x = 10.
args = [1, 2, 3]
func(x, *args)
func2(x, *args)

(1, 2, 3)
1 2 3


10.0

### <font color='blue'>Function stack diagrams and variable scope 

This may be skipped on the first go, but is useful to know going forward. Each function can be thought as having its own data frame - all variables assigned within that frame are "visible" and accessible within the function, but not outside of it. It is said that function variables have limited scope within the function. 

It is possible to define variable to be a global variable and thus be accessible outside the function, but it is advised that the use of global variables is limited to the most necessary cases, because such use may generate unintended errors when global variable is changed erroneously in one part of the code and this error will be propagated throughout the rest of the code. 

A useful tool to visualize function calls, variable values, and variable scope is [python tutor](http://pythontutor.com). 